python/revng/internal/api/rich_manager.py

```diff
@@ -0,0 +1,226 @@
+                sys.stderr.flush()
+
+            for element in old_dir_list:
+                rmtree(element, False)
+
+            with self.lock:
+                credentials = self.credentials
+
+            for hook in self.save_hooks:
+                hook(self.manager, credentials)
+
+        self.shutdown_event.set()
+
+    def submit(self, dir_: str):
+        assert not self.stopping, "Cannot submit while stopping"
+        with self.lock:
+            self.dir_list.append(dir_)
+            self.submit_event.set()
+
+    def set_credentials(self, credentials: str):
+        with self.lock:
+            self.credentials = credentials
+            self.synchronizer.set_credentials(credentials)
+
+
+class ResettableTimer(Thread):
+    def __init__(self, interval: int, func: Callable[[], Any]):
+        super().__init__()
+        self.interval = interval
+        self.func = func
+        self.running = True
+
+        self.lock = Lock()
+        self.next_trigger_time = None
+
+    def run(self):
+        while self.running:
+            if self.next_trigger_time is not None and time.time() > self.next_trigger_time:
+                with self.lock:
+                    self.func()
+                    self.next_trigger_time = None
+            time.sleep(10)
+
+    def reset_countdown(self):
+        with self.lock:
+            self.next_trigger_time = None
+
+    def start_countdown(self):
+        with self.lock:
+            if self.next_trigger_time is None:
+                self.next_trigger_time = time.time() + self.interval
+
+
+class EventType(enum.StrEnum):
+    VOLATILE_CHANGE = enum.auto()
+    SAVE = enum.auto()
+
+
+class RichManager(Manager):
+    SAVE_INTERVAL = 120
+
+    def __init__(
+        self,
+        workdir: str | None = None,
+        flags: Iterable[str] = (),
+        save_hooks: SaveHooks = (),
+    ):
+        synchronizer: Synchronizer | None = None
+        self._storage_saver: StorageSaver | None = None
+
+        if workdir is None:
+            self._temporary_workdir = TemporaryDirectory(prefix="revng-manager-workdir-")
+            workdir = self._temporary_workdir.name
+        elif workdir.startswith("s3://") or workdir.startswith("s3s://"):
+            synchronizer = S3Synchronizer(workdir)
+            self._storage_saver = StorageSaver(self, synchronizer, save_hooks)
+            self._storage_saver.start()
+
+            self._temporary_workdir = TemporaryDirectory(prefix="revng-manager-workdir-")
+            workdir = self._temporary_workdir.name
+
+        if synchronizer is not None:
+            synchronizer.load(Path(workdir))
+
+        self.save_lock = Lock()
+        super().__init__(workdir, flags)
+        self.workdir_path = Path(self.workdir)
+
+        if self._storage_saver is not None:
+            self._wrap_storage_function("save")
+            self._wrap_storage_function("run_analysis")
+            self._wrap_storage_function("run_analyses_list")
```

Comment:
Why are wrapping these functions twice (here and below)?

I also think using inheritance + override is much cleaner than this `setattr` business. It's quite hacky.

Reply:
> A lot of repetition

I'm a bit skeptical that using inheritance is worse than what we have now.
`setattr`, `getattr`, two functions for wrapping declaring a function on the fly...

```python
class RichManager(Manager):

  @override
  def save(...)
    notify_save()
    super().save(...)

  def save(...)
    notify_save()
    super().save(...)

  ...

  def notify_save(self):
    if self._storage_saver is not None:
      ...
```

Pretty simple and intuitive.

> *args, **kwargs

Well... let's not do that. It's literally 5 functions.

> Having extra code because the synchronizer might not be there

Wrap unconditionally and set invoke the synchronizer only if it's there (or make a dummy synchronizer).

Reply:
Managed to use a decorator